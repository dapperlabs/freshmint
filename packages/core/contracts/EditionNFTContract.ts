// @ts-ignore
import * as fcl from '@onflow/fcl';
// @ts-ignore
import * as t from '@onflow/types';

import { NFTContract, CollectionMetadata, prepareCollectionMetadata, Royalty, prepareRoyalties } from './NFTContract';
import { hashMetadata, MetadataMap } from '../metadata';
import { EditionNFTGenerator } from '../generators/EditionNFTGenerator';
import { FreshmintConfig, ContractImports } from '../config';
import { Transaction, TransactionResult, TransactionEvent } from '../transactions';
import { PublicKey, SignatureAlgorithm, HashAlgorithm } from '../crypto';
import { Script } from '../scripts';

export type EditionInput = {
  limit?: number;
  metadata: MetadataMap;
};

export type EditionNFTInput = {
  editionId: string;
  count: number;
};

export type EditionResult = {
  id: string;
  metadata: MetadataMap;
  limit?: number;
};

export type OnChainEdition = {
  id: string;
  size: number;
  limit?: number;
  burned: number;
  isClosed: boolean;
  metadata: { [key: string]: any };
};

export type NFTMintResult = {
  id: string;
  editionId: string;
  editionSerial: string;
  transactionId: string;
};

export class EditionNFTContract extends NFTContract {
  getSource(imports: ContractImports, options?: { saveAdminResourceToContractAccount?: boolean }): string {
    return EditionNFTGenerator.contract({
      imports,
      contractName: this.name,
      schema: this.schema,
      saveAdminResourceToContractAccount: options?.saveAdminResourceToContractAccount,
    });
  }

  deploy({
    publicKey,
    hashAlgorithm,
    collectionMetadata,
    royalties,
    saveAdminResourceToContractAccount,
  }: {
    publicKey: PublicKey;
    hashAlgorithm: HashAlgorithm;
    collectionMetadata: CollectionMetadata;
    royalties?: Royalty[];
    saveAdminResourceToContractAccount?: boolean;
  }): Transaction<string> {
    return new Transaction(
      ({ imports }: FreshmintConfig) => {
        const script = EditionNFTGenerator.deployToNewAccount({ imports });

        const contractCode = this.getSource(imports, { saveAdminResourceToContractAccount });
        const contractCodeHex = Buffer.from(contractCode, 'utf-8').toString('hex');

        const sigAlgo = publicKey.signatureAlgorithm();

        const { royaltyAddresses, royaltyReceiverPaths, royaltyCuts, royaltyDescriptions } = prepareRoyalties(
          royalties ?? [],
        );

        return {
          script,
          args: [
            fcl.arg(this.name, t.String),
            fcl.arg(contractCodeHex, t.String),
            fcl.arg(publicKey.toHex(), t.String),
            fcl.arg(SignatureAlgorithm.toCadence(sigAlgo), t.UInt8),
            fcl.arg(HashAlgorithm.toCadence(hashAlgorithm), t.UInt8),
            fcl.arg(prepareCollectionMetadata(imports.MetadataViews, collectionMetadata), t.Identity),
            fcl.arg(royaltyAddresses, t.Array(t.Address)),
            fcl.arg(royaltyReceiverPaths, t.Array(t.Path)),
            fcl.arg(royaltyCuts, t.Array(t.UFix64)),
            fcl.arg(royaltyDescriptions, t.Array(t.String)),
            fcl.arg(saveAdminResourceToContractAccount ?? false, t.Bool),
          ],
          computeLimit: 9999,
          signers: this.getSigners(),
        };
      },
      ({ events }: TransactionResult) => {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const accountCreatedEvent = events.find((event: TransactionEvent) => event.type === 'flow.AccountCreated')!;

        const address = accountCreatedEvent.data['address'];

        this.setAddress(address);

        return address;
      },
    );
  }

  createEdition(edition: EditionInput): Transaction<EditionResult> {
    const editions = [edition];

    return new Transaction(
      this.makeCreateEditionsTransaction(editions),
      (result) => formatEditionResults(result, editions)[0],
    );
  }

  createEditions(editions: EditionInput[]): Transaction<EditionResult[]> {
    return new Transaction(this.makeCreateEditionsTransaction(editions), (result) =>
      formatEditionResults(result, editions),
    );
  }

  private makeCreateEditionsTransaction(editions: EditionInput[]) {
    return ({ imports }: FreshmintConfig) => {
      const script = EditionNFTGenerator.createEditions({
        imports,
        contractName: this.name,
        contractAddress: this.getAddress(),
        schema: this.schema,
      });

      // Use metadata hash as mint ID
      const mintIds = editions.map((edition) => hashMetadata(this.schema, edition.metadata).toString('hex'));

      const limits = editions.map((edition) => (edition.limit ? edition.limit.toString(10) : undefined));

      return {
        script,
        args: [
          fcl.arg(mintIds, t.Array(t.String)),
          fcl.arg(limits, t.Array(t.Optional(t.UInt64))),
          ...this.schema.fields.map((field) => {
            return fcl.arg(
              editions.map((edition) => field.getValue(edition.metadata)),
              t.Array(field.asCadenceTypeObject()),
            );
          }),
        ],
        computeLimit: 9999,
        signers: this.getSigners(),
      };
    };
  }

  closeEdition(editionId: string): Transaction<void> {
    return new Transaction(({ imports }: FreshmintConfig) => {
      const script = EditionNFTGenerator.closeEdition({
        imports,
        contractName: this.name,
        contractAddress: this.getAddress(),
      });

      return {
        script,
        args: [fcl.arg(editionId, t.UInt64)],
        computeLimit: 9999,
        signers: this.getSigners(),
      };
    }, Transaction.VoidResult);
  }

  mintNFTs({
    editionId,
    count,
    bucket,
  }: {
    editionId: string;
    count: number;
    bucket?: string;
  }): Transaction<NFTMintResult[]> {
    return new Transaction(
      ({ imports }: FreshmintConfig) => {
        const script = EditionNFTGenerator.mint({
          imports,
          contractName: this.name,
          contractAddress: this.getAddress(),
        });

        return {
          script,
          args: [
            fcl.arg(editionId, t.UInt64),
            fcl.arg(count.toString(10), t.Int),
            fcl.arg(bucket, t.Optional(t.String)),
          ],
          computeLimit: 9999,
          signers: this.getSigners(),
        };
      },
      (result) => formatMintResults(result, editionId),
    );
  }

  getEdition(editionId: string): Script<OnChainEdition> {
    const script = EditionNFTGenerator.getEdition({
      contractName: this.name,
      contractAddress: this.getAddress(),
    });

    return new Script(
      () => ({
        script,
        args: (arg, t) => [arg(editionId, t.UInt64)],
        computeLimit: 9999,
      }),
      (result) => ({
        id: result.id,
        size: parseInt(result.size, 10),
        limit: result.limit ? parseInt(result.limit, 10) : undefined,
        burned: parseInt(result.burned, 10),
        isClosed: result.isClosed,
        metadata: result.metadata,
      }),
    );
  }
}

function formatEditionResults({ events }: TransactionResult, editions: EditionInput[]): EditionResult[] {
  const editionEvents = events.filter((event) => event.type.includes('.EditionCreated'));

  return editions.flatMap((edition, i) => {
    const editionEvent: any = editionEvents[i];
    const editionId = editionEvent.data.edition.id;

    return {
      id: editionId,
      metadata: edition.metadata,
      limit: edition.limit,
    };
  });
}

function formatMintResults({ transactionId, events }: TransactionResult, editionId: string): NFTMintResult[] {
  const deposits = events.filter((event) => event.type.includes('.Minted'));

  return deposits.map((deposit) => {
    return {
      id: deposit.data.id,
      editionId,
      editionSerial: deposit.data.serialNumber,
      transactionId,
    };
  });
}
