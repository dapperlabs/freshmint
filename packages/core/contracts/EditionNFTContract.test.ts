import { EditionNFTContract, EditionResult } from './EditionNFTContract';
import { ClaimSaleContract } from './ClaimSaleContract';

import {
  client,
  contractHashAlgorithm,
  contractPublicKey,
  ownerAuthorizer,
  getTestSchema,
  royaltiesTests,
  setupEmulator,
  teardownEmulator,
} from '../testHelpers';

describe('EditionNFTContract', () => {
  beforeAll(setupEmulator);
  afterAll(teardownEmulator);

  const contract = new EditionNFTContract({
    name: 'EditionNFT_Test',
    schema: getTestSchema(false),
    owner: ownerAuthorizer,
  });

  it('should generate a contract', async () => {
    expect(contract.getSource(client.config.imports)).toMatchSnapshot();
  });

  it('should deploy a contract', async () => {
    await client.send(contract.deploy(contractPublicKey, contractHashAlgorithm));
  });

  let edition1: EditionResult;
  let edition2: EditionResult;

  it('should create edition 1', async () => {
    const editionMetadata1 = {
      size: 5,
      metadata: {
        name: 'Edition 1',
        description: 'This is the first edition.',
        thumbnail: 'edition-1.jpeg',
      },
    };

    edition1 = await client.send(contract.createEdition(editionMetadata1));

    const onChainEdition = await client.query(contract.getEdition(edition1.id));

    expect(onChainEdition.id).toEqual(edition1.id);
    expect(onChainEdition.size).toEqual(edition1.size);
    expect(onChainEdition.count).toEqual(0);
  });

  it('should create edition 2', async () => {
    const editionMetadata2 = {
      size: 5,
      metadata: {
        name: 'Edition 2',
        description: 'This is the second edition.',
        thumbnail: 'edition-2.jpeg',
      },
    };

    edition2 = await client.send(contract.createEdition(editionMetadata2));

    const onChainEdition = await client.query(contract.getEdition(edition2.id));

    expect(onChainEdition.id).toEqual(edition2.id);
    expect(onChainEdition.size).toEqual(edition2.size);
    expect(onChainEdition.count).toEqual(0);
  });

  it('should mint 3 edition 1 NFTs into default bucket', async () => {
    const count = 3;

    await client.send(contract.mintNFTs({ editionId: edition1.id, count }));

    const onChainEdition = await client.query(contract.getEdition(edition1.id));

    expect(onChainEdition.id).toEqual(edition1.id);
    expect(onChainEdition.size).toEqual(edition1.size);
    expect(onChainEdition.count).toEqual(count);
  });

  it('should mint remaining 2 edition 1 NFTs into default bucket', async () => {
    const count = 2;

    await client.send(contract.mintNFTs({ editionId: edition1.id, count }));

    const onChainEdition = await client.query(contract.getEdition(edition1.id));

    expect(onChainEdition.id).toEqual(edition1.id);
    expect(onChainEdition.size).toEqual(edition1.size);
    expect(onChainEdition.count).toEqual(edition1.size);
  });

  it('should fail to mint more than edition size', async () => {
    await expect(async () => {
      await client.send(contract.mintNFTs({ editionId: edition1.id, count: 5 }));
    }).rejects.toThrow();

    const onChainEdition = await client.query(contract.getEdition(edition1.id));

    expect(onChainEdition.id).toEqual(edition1.id);
    expect(onChainEdition.size).toEqual(edition1.size);
    expect(onChainEdition.count).toEqual(edition1.size);
  });

  const edition2Bucket = 'edition2';

  it('should mint all edition 2 NFTs into custom bucket', async () => {
    await client.send(contract.mintNFTs({ editionId: edition2.id, count: edition2.size, bucket: edition2Bucket }));
  });

  const sale = new ClaimSaleContract(contract);

  const sale1 = 'sale1';
  const sale2 = 'sale2';

  describe('sale 1', () => {
    it('should start a sale from default bucket', async () => {
      await client.send(sale.start({ id: sale1, price: '10.0' }));
    });

    it('should claim an NFT', async () => {
      await client.send(sale.claimNFT(ownerAuthorizer.address, ownerAuthorizer, sale1));
    });

    it('should stop a sale', async () => {
      await client.send(sale.stop(sale1));
    });
  });

  describe('sale 2', () => {
    it('should start a sale from custom bucket', async () => {
      await client.send(sale.start({ id: sale2, price: '10.0', bucket: edition2Bucket }));
    });

    it('should claim an NFT', async () => {
      await client.send(sale.claimNFT(ownerAuthorizer.address, ownerAuthorizer, sale2));
    });

    it('should stop a sale', async () => {
      await client.send(sale.stop(sale2));
    });
  });

  royaltiesTests(client, contract);
});
