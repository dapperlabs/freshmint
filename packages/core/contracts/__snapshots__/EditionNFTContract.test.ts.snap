// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditionNFTContract should generate a contract 1`] = `
"import NonFungibleToken from 0xf8d6e0586b0a20c7
import MetadataViews from 0xf8d6e0586b0a20c7
import FungibleToken from 0xee82856bf20e2aa6
import FreshmintMetadataViews from 0xf8d6e0586b0a20c7

pub contract EditionNFT_Test: NonFungibleToken {

    pub let version: String

    pub event ContractInitialized()
    pub event Withdraw(id: UInt64, from: Address?)
    pub event Deposit(id: UInt64, to: Address?)
    pub event Minted(id: UInt64, editionID: UInt64, serialNumber: UInt64)
    pub event Burned(id: UInt64)
    pub event EditionCreated(edition: Edition)

    pub let CollectionStoragePath: StoragePath
    pub let CollectionPublicPath: PublicPath
    pub let CollectionPrivatePath: PrivatePath
    pub let AdminStoragePath: StoragePath

    /// The total number of EditionNFT_Test NFTs that have been minted.
    ///
    pub var totalSupply: UInt64

    /// The total number of EditionNFT_Test editions that have been created.
    ///
    pub var totalEditions: UInt64

    /// A list of royalty recipients that is attached to all NFTs
    /// minted by this contract.
    ///
    access(contract) let royalties: [MetadataViews.Royalty]
    
    /// Return the royalty recipients for this contract.
    ///
    pub fun getRoyalties(): [MetadataViews.Royalty] {
        return EditionNFT_Test.royalties
    }

    /// The collection-level metadata for all NFTs minted by this contract.
    ///
    pub let collectionMetadata: MetadataViews.NFTCollectionDisplay

    pub struct Metadata {
    
        pub let name: String
        pub let description: String
        pub let thumbnail: String

        init(
            name: String,
            description: String,
            thumbnail: String,
        ) {
            self.name = name
            self.description = description
            self.thumbnail = thumbnail
        }
    }

    pub struct Edition {

        pub let id: UInt64

        /// The maximum size of this edition.
        ///
        pub let size: UInt64

        /// The number of NFTs minted in this edition.
        ///
        /// The count cannot exceed the edition size.
        ///
        pub var count: UInt64

        /// The metadata for this edition.
        ///
        pub let metadata: Metadata

        init(
            id: UInt64,
            size: UInt64,
            metadata: Metadata
        ) {
            self.id = id
            self.size = size
            self.metadata = metadata

            // An edition starts with a count of zero
            self.count = 0
        }

        /// Increment the NFT count of this edition.
        ///
        /// The count cannot exceed the edition size.
        ///
        access(contract) fun incrementCount() {
            post {
                self.count <= self.size: \\"edition has already reached its maximum size\\"
            }

            self.count = self.count + (1 as UInt64)
        }
    }

    access(self) let editions: {UInt64: Edition}

    pub fun getEdition(id: UInt64): Edition? {
        return EditionNFT_Test.editions[id]
    }

    /// This dictionary indexes editions by their mint ID.
    ///
    /// It is populated at mint time and used to prevent duplicate mints.
    /// The mint ID can be any unique string value,
    /// for example the hash of the edition metadata.
    ///
    access(self) let editionsByMintID: {String: UInt64}

    pub fun getEditionByMintID(mintID: String): UInt64? {
        return EditionNFT_Test.editionsByMintID[mintID]
    }

    pub resource NFT: NonFungibleToken.INFT, MetadataViews.Resolver {

        pub let id: UInt64

        pub let editionID: UInt64
        pub let serialNumber: UInt64

        init(
            editionID: UInt64,
            serialNumber: UInt64
        ) {
            self.id = self.uuid
            self.editionID = editionID
            self.serialNumber = serialNumber
        }

        /// Return the edition that this NFT belongs to.
        ///
        pub fun getEdition(): Edition {
            return EditionNFT_Test.getEdition(id: self.editionID)!
        }

        pub fun getViews(): [Type] {
            return [   
                Type<MetadataViews.NFTView>(),
                Type<MetadataViews.Display>(),
                Type<MetadataViews.ExternalURL>(),
                Type<MetadataViews.NFTCollectionDisplay>(),
                Type<MetadataViews.NFTCollectionData>(),
                Type<MetadataViews.Royalties>(),
                Type<MetadataViews.Edition>()
            ]
        }

        pub fun resolveView(_ view: Type): AnyStruct? {
            let edition = self.getEdition()

            switch view {
                case Type<MetadataViews.NFTView>():
                    return self.resolveNFTView(edition.metadata)
                case Type<MetadataViews.Display>():
                    return self.resolveDisplay(edition.metadata)
                case Type<MetadataViews.ExternalURL>():
                    return self.resolveExternalURL()
                case Type<MetadataViews.NFTCollectionDisplay>():
                    return self.resolveNFTCollectionDisplay()
                case Type<MetadataViews.NFTCollectionData>():
                    return self.resolveNFTCollectionData()
                case Type<MetadataViews.Royalties>():
                    return self.resolveRoyalties()
                case Type<MetadataViews.Edition>():
                    return self.resolveEditionView(edition)
            }

            return nil
        }

        pub fun resolveNFTView(_ metadata: Metadata): MetadataViews.NFTView {
            return MetadataViews.NFTView(
                id: self.id,
                uuid: self.uuid,
                display: self.resolveDisplay(metadata),
                externalURL: self.resolveExternalURL(),
                collectionData: self.resolveNFTCollectionData(),
                collectionDisplay: self.resolveNFTCollectionDisplay(),
                royalties : self.resolveRoyalties(),
                traits: nil
            )
        }
        
        pub fun resolveDisplay(_ metadata: Metadata): MetadataViews.Display {
            return MetadataViews.Display(
                name: metadata.name,
                description: metadata.description,
                thumbnail: FreshmintMetadataViews.ipfsFile(file: metadata.thumbnail)
            )
        }
        
        pub fun resolveExternalURL(): MetadataViews.ExternalURL {
            let collectionURL = EditionNFT_Test.collectionMetadata.externalURL.url
            let nftOwner = self.owner!.address.toString()
            let nftID = self.id.toString()
            return MetadataViews.ExternalURL(collectionURL.concat(\\"/nfts/\\").concat(nftOwner).concat(\\"/\\").concat(nftID))
        }
        
        pub fun resolveNFTCollectionDisplay(): MetadataViews.NFTCollectionDisplay {
            return EditionNFT_Test.collectionMetadata
        }
        
        pub fun resolveNFTCollectionData(): MetadataViews.NFTCollectionData {
            return MetadataViews.NFTCollectionData(
                storagePath: EditionNFT_Test.CollectionStoragePath,
                publicPath: EditionNFT_Test.CollectionPublicPath,
                providerPath: EditionNFT_Test.CollectionPrivatePath,
                publicCollection: Type<&EditionNFT_Test.Collection{EditionNFT_Test.EditionNFT_TestCollectionPublic}>(),
                publicLinkedType: Type<&EditionNFT_Test.Collection{EditionNFT_Test.EditionNFT_TestCollectionPublic, NonFungibleToken.CollectionPublic, NonFungibleToken.Receiver, MetadataViews.ResolverCollection}>(),
                providerLinkedType: Type<&EditionNFT_Test.Collection{EditionNFT_Test.EditionNFT_TestCollectionPublic, NonFungibleToken.CollectionPublic, NonFungibleToken.Provider, MetadataViews.ResolverCollection}>(),
                createEmptyCollectionFunction: (fun (): @NonFungibleToken.Collection {
                    return <-EditionNFT_Test.createEmptyCollection()
                })
            )
        }
        
        pub fun resolveRoyalties(): MetadataViews.Royalties {
            return MetadataViews.Royalties(EditionNFT_Test.getRoyalties())
        }
        
        pub fun resolveEditionView(_ edition: Edition): MetadataViews.Edition {
            return MetadataViews.Edition(
                name: \\"Edition\\",
                number: self.serialNumber,
                max: edition.size
            )
        }

        destroy() {
            EditionNFT_Test.totalSupply = EditionNFT_Test.totalSupply - (1 as UInt64)

            emit Burned(id: self.id)
        }
    }

    pub resource interface EditionNFT_TestCollectionPublic {
        pub fun deposit(token: @NonFungibleToken.NFT)
        pub fun getIDs(): [UInt64]
        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT
        pub fun borrowEditionNFT_Test(id: UInt64): &EditionNFT_Test.NFT? {
            post {
                (result == nil) || (result?.id == id):
                    \\"Cannot borrow EditionNFT_Test reference: The ID of the returned reference is incorrect\\"
            }
        }
    }
    
    pub resource Collection: EditionNFT_TestCollectionPublic, NonFungibleToken.Provider, NonFungibleToken.Receiver, NonFungibleToken.CollectionPublic, MetadataViews.ResolverCollection {
        
        /// A dictionary of all NFTs in this collection indexed by ID.
        ///
        pub var ownedNFTs: @{UInt64: NonFungibleToken.NFT}
    
        init () {
            self.ownedNFTs <- {}
        }
    
        /// Remove an NFT from the collection and move it to the caller.
        ///
        pub fun withdraw(withdrawID: UInt64): @NonFungibleToken.NFT {
            let token <- self.ownedNFTs.remove(key: withdrawID) 
                ?? panic(\\"Requested NFT to withdraw does not exist in this collection\\")
    
            emit Withdraw(id: token.id, from: self.owner?.address)
    
            return <- token
        }
    
        /// Deposit an NFT into this collection.
        ///
        pub fun deposit(token: @NonFungibleToken.NFT) {
            let token <- token as! @EditionNFT_Test.NFT
    
            let id: UInt64 = token.id
    
            // add the new token to the dictionary which removes the old one
            let oldToken <- self.ownedNFTs[id] <- token
    
            emit Deposit(id: id, to: self.owner?.address)
    
            destroy oldToken
        }
    
        /// Return an array of the NFT IDs in this collection.
        ///
        pub fun getIDs(): [UInt64] {
            return self.ownedNFTs.keys
        }
    
        /// Return a reference to an NFT in this collection.
        ///
        /// This function panics if the NFT does not exist in this collection.
        ///
        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT {
            return (&self.ownedNFTs[id] as &NonFungibleToken.NFT?)!
        }
    
        /// Return a reference to an NFT in this collection
        /// typed as EditionNFT_Test.NFT.
        ///
        /// This function returns nil if the NFT does not exist in this collection.
        ///
        pub fun borrowEditionNFT_Test(id: UInt64): &EditionNFT_Test.NFT? {
            if self.ownedNFTs[id] != nil {
                let ref = (&self.ownedNFTs[id] as auth &NonFungibleToken.NFT?)!
                return ref as! &EditionNFT_Test.NFT
            }
    
            return nil
        }
    
        /// Return a reference to an NFT in this collection
        /// typed as MetadataViews.Resolver.
        ///
        /// This function panics if the NFT does not exist in this collection.
        ///
        pub fun borrowViewResolver(id: UInt64): &AnyResource{MetadataViews.Resolver} {
            let nft = (&self.ownedNFTs[id] as auth &NonFungibleToken.NFT?)!
            let nftRef = nft as! &EditionNFT_Test.NFT
            return nftRef as &AnyResource{MetadataViews.Resolver}
        }
    
        destroy() {
            destroy self.ownedNFTs
        }
    }
    
    /// Return a new empty collection.
    ///
    pub fun createEmptyCollection(): @NonFungibleToken.Collection {
        return <- create Collection()
    }

    /// The administrator resource used to mint and reveal NFTs.
    ///
    pub resource Admin {

        /// Create a new NFT edition.
        ///
        /// This function does not mint any NFTs. It only creates the
        /// edition data that will later be associated with minted NFTs.
        ///
        pub fun createEdition(
            mintID: String,
            size: UInt64,
            name: String,
            description: String,
            thumbnail: String,
        ): UInt64 {
            let metadata = Metadata(
                name: name,
                description: description,
                thumbnail: thumbnail,
            )

            // Prevent multiple editions from being minted with the same mint ID
            assert(
                EditionNFT_Test.editionsByMintID[mintID] == nil,
                message: \\"an edition has already been created with mintID=\\".concat(mintID)
            )

            let edition = Edition(
                id: EditionNFT_Test.totalEditions,
                size: size,
                metadata: metadata
            )

            // Save the edition
            EditionNFT_Test.editions[edition.id] = edition

            // Update the mint ID index
            EditionNFT_Test.editionsByMintID[mintID] = edition.id

            emit EditionCreated(edition: edition)

            EditionNFT_Test.totalEditions = EditionNFT_Test.totalEditions + (1 as UInt64)

            return edition.id
        }

        /// Mint a new NFT.
        ///
        /// This function will mint the next NFT in this edition
        /// and automatically assign the serial number.
        ///
        /// This function will panic if the edition has already
        /// reached its maximum size.
        ///
        pub fun mintNFT(editionID: UInt64): @EditionNFT_Test.NFT {
            let edition = EditionNFT_Test.editions[editionID]
                ?? panic(\\"edition does not exist\\")

            // Increase the edition count by one
            edition.incrementCount()

            // The NFT serial number is the new edition count
            let serialNumber = edition.count

            let nft <- create EditionNFT_Test.NFT(
                editionID: editionID,
                serialNumber: serialNumber
            )

            // Save the updated edition
            EditionNFT_Test.editions[editionID] = edition

            emit Minted(id: nft.id, editionID: editionID, serialNumber: serialNumber)

            EditionNFT_Test.totalSupply = EditionNFT_Test.totalSupply + (1 as UInt64)

            return <- nft
        }
    }

    /// Return a public path that is scoped to this contract.
    ///
    pub fun getPublicPath(suffix: String): PublicPath {
        return PublicPath(identifier: \\"EditionNFT_Test_\\".concat(suffix))!
    }

    /// Return a private path that is scoped to this contract.
    ///
    pub fun getPrivatePath(suffix: String): PrivatePath {
        return PrivatePath(identifier: \\"EditionNFT_Test_\\".concat(suffix))!
    }

    /// Return a storage path that is scoped to this contract.
    ///
    pub fun getStoragePath(suffix: String): StoragePath {
        return StoragePath(identifier: \\"EditionNFT_Test_\\".concat(suffix))!
    }

    /// Return a collection name with an optional bucket suffix.
    ///
    pub fun makeCollectionName(bucketName maybeBucketName: String?): String {
        if let bucketName = maybeBucketName {
            return \\"Collection_\\".concat(bucketName)
        }

        return \\"Collection\\"
    }

    /// Return a queue name with an optional bucket suffix.
    ///
    pub fun makeQueueName(bucketName maybeBucketName: String?): String {
        if let bucketName = maybeBucketName {
            return \\"Queue_\\".concat(bucketName)
        }

        return \\"Queue\\"
    }

    priv fun initAdmin(admin: AuthAccount) {
        // Create an empty collection and save it to storage
        let collection <- EditionNFT_Test.createEmptyCollection()

        admin.save(<- collection, to: EditionNFT_Test.CollectionStoragePath)

        admin.link<&EditionNFT_Test.Collection>(EditionNFT_Test.CollectionPrivatePath, target: EditionNFT_Test.CollectionStoragePath)

        admin.link<&EditionNFT_Test.Collection{NonFungibleToken.CollectionPublic, EditionNFT_Test.EditionNFT_TestCollectionPublic, MetadataViews.ResolverCollection}>(EditionNFT_Test.CollectionPublicPath, target: EditionNFT_Test.CollectionStoragePath)
        
        // Create an admin resource and save it to storage
        let adminResource <- create Admin()

        admin.save(<- adminResource, to: self.AdminStoragePath)
    }

    init(collectionMetadata: MetadataViews.NFTCollectionDisplay, royalties: [MetadataViews.Royalty], admin: AuthAccount) {

        self.version = \\"0.3.0\\"

        self.CollectionPublicPath = EditionNFT_Test.getPublicPath(suffix: \\"Collection\\")
        self.CollectionStoragePath = EditionNFT_Test.getStoragePath(suffix: \\"Collection\\")
        self.CollectionPrivatePath = EditionNFT_Test.getPrivatePath(suffix: \\"Collection\\")

        self.AdminStoragePath = EditionNFT_Test.getStoragePath(suffix: \\"Admin\\")

        self.royalties = royalties
        self.collectionMetadata = collectionMetadata

        self.totalSupply = 0
        self.totalEditions = 0

        self.editions = {}
        self.editionsByMintID = {}
        
        self.initAdmin(admin: admin)

        emit ContractInitialized()
    }
}
"
`;
